### PR TITLE
Dashboard stats: replace XLPagerTabStrip with WPiOS implementation

### PR DESCRIPTION
**Pull request by jaclync**
Part of #1911 

## Why

In the dashboard, we currently use a third-party library `XLPagerTabStrip` for the tabbed stats where there is a tab bar for selecting the time range with scrollable stats below it. There have been discussions of removing `XLPagerTabStrip` from our app as linked in #1911, and we reached a conclusion to replace it with the WPiOS `FilterTabBar` implementation in p91TBi-47N-p2 after design confirmation. This used to be a tech debt item left for the future. Now that we are moving to one single vertical `UIScrollView` to account for the dynamic onboarding card height, the use of `XLPagerTabStrip` causes a bunch of issues beyond our control from my testing. Replacing `XLPagerTabStrip` enables us to make changes that suit our design. [The library](https://github.com/xmartlabs/XLPagerTabStrip) has also accumulated 250+ open issues and could go out of maintenance anytime. Since WPiOS has been using the `FilterTabBar` + `TabbedViewController` implementation for a few years, we have a pretty stable alternative to replace with.

## Description

Apology for the big diffs, the new diffs are mostly from the WPiOS [`FilterTabBar`](https://github.com/wordpress-mobile/WordPress-iOS/blob/c67f3b6205c1bc40085b390ec5e46faf5e281df9/WordPress/Classes/ViewRelated/System/FilterTabBar.swift) and  [`TabbedViewController`](https://github.com/wordpress-mobile/WordPress-iOS/blob/c67f3b6205c1bc40085b390ec5e46faf5e281df9/WordPress/Classes/ViewRelated/Reader/Manage/TabbedViewController.swift) with some updates to fit our use case like [updates to make the build pass](https://github.com/woocommerce/woocommerce-ios/pull/8942/commits/d01d6b23bb2c677a7ec790d591ce3ae2f0b84589) and [programmatically selecting a time range tab](https://github.com/woocommerce/woocommerce-ios/pull/8942/commits/c68b19c4b1c8a60c48dcd5fd3a8c1f5728f2b65e).

Other than the new classes adapted from WPiOS, the majority of the diffs are on `StoreStatsAndTopPerformersViewController` where it's now a subclass of `TabbedViewController` instead of `XLPagerTabStrip.ButtonBarPagerTabStripViewController`. `selection: Int` from `TabbedViewController` is the source of truth of the selected tab index as shown in the UI, and I kept the observable version of it `selectedTimeRangeIndex` to minimize the changes. I spent quite some time [updating the tab bar button styles](https://github.com/woocommerce/woocommerce-ios/pull/8942/commits/ed794334f81c9fd9849aaf7adb8d3f2aadfb7824) in order to remove the iOS 15+ build warnings on `contentEdgeInsets` because the new button configuration breaks certain button style customizations with the old APIs.

⚠️ One style that I couldn't get exactly right is the spacing of the tab bar buttons. When setting the `UIStackView`'s `spacing` to `0` and `distribution` to `euqalSpacing`, a ~16px spacing is still shown. From the [Apple documentation](https://developer.apple.com/documentation/uikit/uistackview/1616225-spacing) on `UIStackView.spacing`, it's the minimum spacing but I ran out of ideas on how to set the actual spacing and I set a negative value as a workaround. Please feel free to share any ideas on this!

## Testing instructions

- Log in if needed --> the stats should look like before, with the same functionalities except for horizontal scrolling between time range tabs
- Rotate the device --> the dashboard should work as before in the landscape mode
- Tap another time range tab --> the corresponding time range stats should be shown
- Relaunch the app --> the last selected tab and stats should be shown
- Switch to another store --> the stats for the new store should be shown

## Screenshots

\ | Before | After 
-- | -- | --
light | Simulator Screen Shot - iPhone 14 Pro - 2023-02-23 at 16 09 51 | Simulator Screen Shot - iPhone 14 - 2023-02-23 at 16 09 53
dark | Simulator Screen Shot - iPhone 14 Pro - 2023-02-23 at 16 11 31 | Simulator Screen Shot - iPhone 14 - 2023-02-23 at 16 11 33

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/FilterTabBar/TabbedViewController.swift [hunk 1]
```diff
@@ -0,0 +1,105 @@
+import UIKit
+
+// NOTE: this file is adapted from WPiOS.
```
Comment (mokagio) [inline WooCommerce/Classes/ViewRelated/FilterTabBar/TabbedViewController.swift]: What do you think of adding a link to the source version?

```suggestion
// NOTE: this file is adapted from WPiOS.
//
// https://github.com/wordpress-mobile/WordPress-iOS/blob/67a5d840eeb9ac027404ad71ed246c645de33687/WordPress/Classes/ViewRelated/Reader/Manage/TabbedViewController.swift
```

File: WooCommerce/Classes/ViewRelated/FilterTabBar/FilterTabBar.swift [hunk 2]
```diff
@@ -0,0 +1,546 @@
+import UIKit
+
+// NOTE: this file is adapted from WPiOS.
```
Comment (mokagio) [inline WooCommerce/Classes/ViewRelated/FilterTabBar/FilterTabBar.swift]: What do you think of adding a link to the source version?

```suggestion
// NOTE: this file is adapted from WPiOS.
//
// https://github.com/wordpress-mobile/WordPress-iOS/blob/67a5d840eeb9ac027404ad71ed246c645de33687/WordPress/Classes/ViewRelated/System/FilterTabBar.swift
```

**Review by mokagio — COMMENTED** [review]
Did you consider extracting `FilterTabBar` and `TabbedViewController` in a Swift Package?

Duplicating the code has the undoubted advantage of making customization easier as well as making the development process faster, so I'm not necessarily arguing against it.

Just curious if that's an option you considered and rejected, or if it's something you'd consider as a future improvement. I.e. make it work well here, extract it in a Swift Package, and finally adopt the package in Jetpack/WordPress iOS.

**Comment by jaclync** [comment]
> Did you consider extracting `FilterTabBar` and `TabbedViewController` in a Swift Package?
> 
> Duplicating the code has the undoubted advantage of making customization easier as well as making the development process faster, so I'm not necessarily arguing against it.
> 
> Just curious if that's an option you considered and rejected, or if it's something you'd consider as a future improvement. I.e. make it work well here, extract it in a Swift Package, and finally adopt the package in Jetpack/WordPress iOS.

Hey @mokagio, great question, and thanks for taking a look! I've considered adding this to the [WordPressUI library](https://github.com/wordpress-mobile/WordPressUI-iOS), but it will likely take a few days of back and forth and to also test the changes in WPiOS while we still have the priority of shipping features. There are also some changes to be made to work in both WPiOS and WCiOS. Moving it to a Swift Package could be more light weight but becomes separated from the central UI components library. I hope we can still make the move when we have some cooldown time. WDYT about the location, WordPressUI library or a new Swift Pcakage?

**Review by selanthiraiyan — APPROVED** [review]
This works as expected. 🚢

**Comment by mokagio** [comment]
@jaclync thank you for the context.

I agree that extracting the components _somewhere_ would have added a lot of unnecessary friction to the feature development. 👍 

I hadn't considered WordPressUI, actually. 😳 It's been a while since I landed on that repo 😅 Thank you for the reminder.

I think it'd likely be best to move the views there, then. Not that they wouldn't make sense standalone, but realistically we are not looking at building a tabbed screen for the community, only having something tailored for us. With that in mind, it makes sense for it to live in the general purpose UI component library.

File: WooCommerce/Classes/ViewRelated/FilterTabBar/TabbedViewController.swift (hunk 1 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/FilterTabBar/TabbedViewController.swift]: Good point! added in https://github.com/woocommerce/woocommerce-ios/pull/8942/commits/27adc006d64aaef1db315022feb695c18bb9c640

File: WooCommerce/Classes/ViewRelated/FilterTabBar/FilterTabBar.swift (hunk 2 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/FilterTabBar/FilterTabBar.swift]: Added in https://github.com/woocommerce/woocommerce-ios/pull/8942/commits/27adc006d64aaef1db315022feb695c18bb9c640 - the file name is too long so I had to break it into two lines.

**Comment by jaclync** [comment]
> I agree that extracting the components _somewhere_ would have added a lot of unnecessary friction to the feature development. 👍
> 
> I hadn't considered WordPressUI, actually. 😳 It's been a while since I landed on that repo 😅 Thank you for the reminder.
> 
> I think it'd likely be best to move the views there, then. Not that they wouldn't make sense standalone, but realistically we are not looking at building a tabbed screen for the community, only having something tailored for us. With that in mind, it makes sense for it to live in the general purpose UI component library.

Thanks for understanding and for your input, I agree that extracting them to WordPressUI makes the most sense. I created an issue https://github.com/woocommerce/woocommerce-ios/issues/8975 to track this for us to address in the future.

**Pull request merged**